src/emqx_shared_sub.erl

```diff
@@ -264,23 +267,31 @@ pick(sticky, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
             {fresh, Sub0};
         false ->
             %% randomly pick one for the first message
-            {Type, Sub} = do_pick(random, ClientId, SourceTopic, Group, Topic, [Sub0 | FailedSubs]),
-            %% stick to whatever pick result
-            erlang:put({shared_sub_sticky, Group, Topic}, Sub),
-            {Type, Sub}
+            FailedSubs1 = maps_put_new(FailedSubs, Sub0, inactive),
+						case do_pick(random, ClientId, SourceTopic, Group, Topic, FailedSubs1) of
+              false -> false;
+              {Type, Sub} ->
+                %% stick to whatever pick result
+                erlang:put({shared_sub_sticky, Group, Topic}, Sub),
+                {Type, Sub}
+            end
     end;
 pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
     do_pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs).
 
 do_pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
     All = subscribers(Group, Topic),
-    case All -- FailedSubs of
+    case lists:filter(fun(Sub) -> not maps:is_key(Sub, FailedSubs) end, All) of
         [] when All =:= [] ->
             %% Genuinely no subscriber
             false;
         [] ->
             %% All offline? pick one anyway
-            {retry, pick_subscriber(Group, Topic, Strategy, ClientId, SourceTopic, All)};
+            %% We redispatch only to subs who dropped the message because inflight was full.
+            Dropped = maps_find_by(FailedSubs, fun({SubPid, FailReason}) ->
```

Comment:
It seems changed the original logic. How about if all FailReason are not dropped (i.e: no_connection)? 🤔

Reply:
Perhaps we need a new function to express how we can re-pick a suitable subscriber when all of them do not want to deliver the message.

Reply:
`no_connection` is a valid nack, we shouldn't deliver message to disconnected sub. But when there are subs who have their inflights full, we should deliver to them

Reply:
Maybe we can add some logging? Because since no subscribers were found that could be delivered, this would cause messages to be dropped

Reply:
I think we can leave the previous behaviour, where some subscriber is picked anyway.